### PR TITLE
Extract authorization information for subgraph queries

### PR DESCRIPTION
**Pull request by Geal**
Query deduplication was already taking authorization information into account in its key, but that was for the global authorization context, ie the intersection of what the query authorization requires, and what the request token provides.
This was very coarse grained because we could have some subgraph queries with different authorization requirements, or even no authorization requirements.
This PR extracts the authorization info from subgraph queries, and intersects that with the request's authorization context, to make more fine grained query deduplication.
The entire process happens during query planning, and the relevant information is stored in the Fetch plan node.

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: apollo-router/src/plugins/authorization/policy.rs [hunk 1]
```diff
@@ -61,6 +67,38 @@ impl<'a> PolicyExtractionVisitor<'a> {
...
+    fn entities_operation(&mut self, node: &ast::OperationDefinition) -> Result<(), BoxError> {
+        use crate::spec::query::traverse::Visitor;
+
+        if node.selection_set.len() != 1 {
+            return Err("invalid number of selections for _entities query".into());
+        }
+
+        match node.selection_set.first() {
+            Some(ast::Selection::Field(field)) => {
+                if field.name.as_str() != "_entities" {
```
Comment (bnjjj) [inline apollo-router/src/plugins/authorization/policy.rs:80]: use the constant for `_entities`

File: apollo-router/src/plugins/authorization/scopes.rs [hunk 2]
```diff
@@ -61,6 +67,38 @@ impl<'a> ScopeExtractionVisitor<'a> {
...
+    fn entities_operation(&mut self, node: &ast::OperationDefinition) -> Result<(), BoxError> {
+        use crate::spec::query::traverse::Visitor;
+
+        if node.selection_set.len() != 1 {
+            return Err("invalid number of selections for _entities query".into());
+        }
+
+        match node.selection_set.first() {
+            Some(ast::Selection::Field(field)) => {
+                if field.name.as_str() != "_entities" {
```
Comment (bnjjj) [inline apollo-router/src/plugins/authorization/scopes.rs:80]: use the constant for `_entities`

File: apollo-router/src/query_planner/fetch.rs [hunk 3]
```diff
@@ -454,4 +462,21 @@ impl FetchNode {
     pub(crate) fn operation_kind(&self) -> &OperationKind {
         &self.operation_kind
     }
+
+    pub(crate) fn extract_authorization_metadata(
+        &mut self,
+        schema: &apollo_compiler::Schema,
+        global_authorisation_cache_key: &CacheKeyMetadata,
+    ) {
+        let doc = Document::parse(&self.operation, "query.graphql");
```
Comment (bnjjj) [inline apollo-router/src/query_planner/fetch.rs:471]: Does it mean we are reparsing something ?

Comment (Geal) [reply apollo-router/src/query_planner/fetch.rs:472]: we are parsing the subgraph query, it was not parsed before

**Review by o0Ignition0o — APPROVED** [review]

**Pull request merged**